File: 210-course-schedule-ii/solution.ts
function findOrder(numCourses: number, prerequisites: number[][]): number[] {
  const adj = Array.from({ length: numCourses }).map(() => [] as boolean[]);
  const ref = Array.from({ length: numCourses }).map(() => 0);
  for (let [i, j] of prerequisites) {
    adj[j][i] = true;
    ref[i]++;
  }

  const order: number[] = [];
  for (let i = 0; i < numCourses; ++i) {
    let node = 0;
    while (node < numCourses && ref[node] !== 0) ++node;
    if (node == numCourses) return [];

    ref[node] = -1;
    order.push(node);
    for (let j = 0; j < numCourses; ++j) {
      if (adj[node][j]) ref[j]--;
    }
  }
  return order;
}

File: 210-course-schedule-ii/solution.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> e(numCourses);
        vector<int> indeg(numCourses, 0);
        for(auto& x : prerequisites) {
            e[x[1]].push_back(x[0]);
            indeg[x[0]]++;
        }
        queue<int> qu;
        for(int i = 0; i < numCourses; i++) {
            if(indeg[i] == 0) qu.push(i);
        }
        vector<int> order;
        while(!qu.empty()) {
            auto u = qu.front();
            qu.pop();
            order.push_back(u);
            for(auto& v : e[u]) {
                if(--indeg[v] == 0) qu.push(v);
            }
        }
        return (order.size() == numCourses) ? order : vector<int>({});
    }
};
